lib/datadog/tracing/trace_segment.rb

```diff
@@ -42,18 +42,24 @@ def initialize(
 
         # Set well-known tags
         self.agent_sample_rate = agent_sample_rate
-        self.hostname = (hostname && hostname.dup)
-        self.lang = (lang && lang.dup)
-        self.name = (name && name.dup)
-        self.origin = (origin && origin.dup)
+        self.hostname = hostname
+        self.lang = lang
+        self.name = (name.frozen? ? name : name.dup)
+        self.origin = (origin.frozen? ? origin : origin.dup)
         self.process_id = process_id
         self.rate_limiter_rate = rate_limiter_rate
-        self.resource = (resource && resource.dup)
+        self.resource = (resource.frozen? ? resource : resource.dup)
         self.rule_sample_rate = rule_sample_rate
-        self.runtime_id = (runtime_id && runtime_id.dup)
+        self.runtime_id = runtime_id
         self.sample_rate = sample_rate
         self.sampling_priority = sampling_priority
-        self.service = (service && service.dup)
+        self.service = (service.frozen? ? service : service.dup)
+      end
+
+      if RUBY_VERSION < '2.2' # nil.dup only fails in Ruby 2.1
+        def nil.dup
+          self
+        end
       end
```

Comment:
I'm proposing we keep the desired pattern for newer Rubies, at the expense of Ruby 2.1 (having this custom block for 2.1).

It's easy for us to clean this up when 2.1 support is dropped, and leaves the performance improvements for all supported Rubies.

Adding an additional check for `nil` in each of the `(x.frozen? ? x : x.dup)` lines will defeat the performance improvements, reverting us back to square one.

Reply:
Agree with improving perf for Ruby > 2.1 at the expense of 2.1.